### ISSUE TITLE
zpool scrub only scans the special vdev

### ISSUE THREAD

**Issue by speed47** [openzfs/zfs#13706]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Ubuntu
Distribution Version	| 22.04
Kernel Version	| 5.15.53 (vanilla)
Architecture	| x86_64
OpenZFS Version	| 2.0.0 - 2.1.99 (commit 34aa0f0487705671c81262adb7646a90d15c5a12)

### Describe the problem you're observing
When running a `zpool scrub` on a zpool with 2 (data) mirrors with HDDs and one special mirror with SSDs, only the special mirror seems to be scanned, as can be seen below:

```
# zpool status z
  pool: z
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
        The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(7) for details.
  scan: scrub in progress since Fri Jul 29 19:30:18 2022
        817G scanned at 51.1G/s, 817G issued at 51.1G/s, 12.8T total
        0B repaired, 6.21% done, 00:04:01 to go
remove: Removal of vdev 7 copied 38.6G in 0h7m, completed on Fri Jul 29 11:01:51 2022
        2.11M memory used for removed device mappings
config: 

        NAME                                           STATE     READ WRITE CKSUM
        z                                              ONLINE       0     0     0
          mirror-0                                     ONLINE       0     0     0
            usb-WD_Elements_25A3_35514747344C4242-0:0  ONLINE       0     0     0
            ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F         ONLINE       0     0     0
          mirror-1                                     ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF   ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR   ONLINE       0     0     0
        special
          mirror-9                                     ONLINE       0     0     0
            vgssd1000/z_special_ssd1000                ONLINE       0     0     0
            vg/z_special_ssd300                        ONLINE       0     0     0
```

The status line `817G scanned at 51.1G/s, 817G issued at 51.1G/s, 12.8T total` progresses very fast, and the amount scanned and issued are always exactly the same, up to 12.8T after less than 5 minutes when the scrub is considered complete, and obviously none of the HDDs have actually scanned their blocks.
This can also be confirmed running `iostat -x 1` in parallel: only the SSD disks are actively working, the others are mostly idle.

### Describe how to reproduce the problem
This is reproducible 100% of the time, at least with my zpool setup: I have no way to actually scrub the data (AFAICT)
The setup details can be found below. It should be noted that there were a few add/remove/attach/detach operations on the special vdevs, which implies that there are several indirect vdevs (as can be seen in the `zdb -C` dump).
The pool has been created under v2.1.5 with `-o ashift=12 -o feature@encryption=enabled`

<details>
  <summary>zpool list -v</summary>

```
  NAME                                            SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
z                                              13.7T  12.8T   893G        -         -     0%    93%  1.00x    ONLINE  -
  mirror-0                                     10.9T  10.1T   805G        -         -     0%  92.8%      -    ONLINE
    usb-WD_Elements_25A3_35514747344C4242-0:0  10.9T      -      -        -         -      -      -      -    ONLINE
    ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F         10.9T      -      -        -         -      -      -      -    ONLINE
  mirror-1                                     2.72T  2.69T  27.7G        -         -     0%  99.0%      -    ONLINE
    ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF   2.73T      -      -        -         -      -      -      -    ONLINE
    ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR   2.73T      -      -        -         -      -      -      -    ONLINE
special                                            -      -      -        -         -      -      -      -         -
  mirror-9                                     99.5G  39.0G  60.5G        -         -    10%  39.2%      -    ONLINE
    vgssd1000/z_special_ssd1000                 100G      -      -        -         -      -      -      -    ONLINE
    vg/z_special_ssd300                         100G      -      -        -         -      -      -      -    ONLINE
```

</details>

<details>
<summary>zdb -C</summary>

```
MOS Configuration:
        version: 5000
        name: 'z'
        state: 0
        txg: 191183
        pool_guid: 17784954522390314791
        errata: 0
        hostname: 'nas'
        com.delphix:has_per_vdev_zaps
        vdev_children: 10
        vdev_tree:
            type: 'root'
            id: 0
            guid: 17784954522390314791
            create_txg: 4
            children[0]:
                type: 'mirror'
                id: 0
                guid: 2972948471517465359
                metaslab_array: 65
                metaslab_shift: 34
                ashift: 12
                asize: 12000089866240
                is_log: 0
                create_txg: 4
                com.delphix:vdev_zap_top: 129
                children[0]:
                    type: 'disk'
                    id: 0
                    guid: 2886530369459906751
                    path: '/dev/disk/by-id/usb-WD_Elements_25A3_35514747344C4242-0:0-part1'
                    devid: 'usb-WD_Elements_25A3_35514747344C4242-0:0-part1'
                    phys_path: 'pci-0000:00:14.0-usb-0:4:1.0-scsi-0:0:0:0'
                    whole_disk: 1
                    DTL: 800
                    create_txg: 4
                    com.delphix:vdev_zap_leaf: 130
                children[1]:
                    type: 'disk'
                    id: 1
                    guid: 4718213307199790485
                    path: '/dev/disk/by-id/ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F-part1'
                    devid: 'ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F-part1'
                    phys_path: 'pci-0000:00:1f.2-ata-5.0'
                    whole_disk: 1
                    DTL: 799
                    create_txg: 4
                    com.delphix:vdev_zap_leaf: 131
            children[1]:
                type: 'mirror'
                id: 1
                guid: 1012032482718870771
                metaslab_array: 79
                metaslab_shift: 34
                ashift: 12
                asize: 3000578342912
                is_log: 0
                create_txg: 13
                com.delphix:vdev_zap_top: 385
                children[0]:
                    type: 'disk'
                    id: 0
                    guid: 9926059460638940645
                    path: '/dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF-part1'
                    devid: 'ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF-part1'
                    phys_path: 'pci-0000:00:14.0-usb-0:3:1.0-scsi-0:0:0:0'
                    whole_disk: 1
                    DTL: 798
                    create_txg: 13
                    com.delphix:vdev_zap_leaf: 386
                children[1]:
                    type: 'disk'
                    id: 1
                    guid: 1153354702596158308
                    path: '/dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR-part1'
                    devid: 'ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR-part1'
                    phys_path: 'pci-0000:00:14.0-usb-0:3:1.0-scsi-0:0:0:1'
                    whole_disk: 1
                    DTL: 797
                    create_txg: 13
                    com.delphix:vdev_zap_leaf: 387
            children[2]:
                type: 'indirect'
                id: 2
                guid: 6093511207376287821
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 21470117888
                is_log: 0
                com.delphix:indirect_object: 73129
                com.delphix:indirect_births: 73131
                create_txg: 37
                com.delphix:vdev_zap_top: 390
            children[3]:
                type: 'indirect'
                id: 3
                guid: 17247912709748369651
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 21470117888
                is_log: 0
                com.delphix:indirect_object: 72946
                com.delphix:indirect_births: 72948
                com.delphix:prev_indirect_vdev: 2
                create_txg: 37
                com.delphix:vdev_zap_top: 392
            children[4]:
                type: 'indirect'
                id: 4
                guid: 17978966862030276124
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 48313663488
                is_log: 0
                com.delphix:indirect_object: 73354
                com.delphix:indirect_births: 73356
                com.delphix:prev_indirect_vdev: 3
                create_txg: 72529
                com.delphix:vdev_zap_top: 72667
            children[5]:
                type: 'indirect'
                id: 5
                guid: 3364556388247222535
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 42944954368
                is_log: 0
                non_allocating: 1
                com.delphix:indirect_object: 1457
                com.delphix:indirect_births: 1460
                com.delphix:prev_indirect_vdev: 4
                create_txg: 72694
                com.delphix:vdev_zap_top: 73085
            children[6]:
                type: 'indirect'
                id: 6
                guid: 9730150406886455653
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 53682372608
                is_log: 0
                non_allocating: 1
                com.delphix:indirect_object: 6134
                com.delphix:indirect_births: 6136
                com.delphix:prev_indirect_vdev: 5
                create_txg: 176839
                com.delphix:vdev_zap_top: 675
            children[7]:
                type: 'indirect'
                id: 7
                guid: 10676060885641407609
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 107369463808
                is_log: 0
                non_allocating: 1
                com.delphix:indirect_object: 7466
                com.delphix:indirect_births: 7468
                com.delphix:prev_indirect_vdev: 8
                create_txg: 184452
                com.delphix:vdev_zap_top: 6124
            children[8]:
                type: 'indirect'
                id: 8
                guid: 11608273382032491262
                whole_disk: 0
                metaslab_array: 0
                metaslab_shift: 29
                ashift: 12
                asize: 53682372608
                is_log: 0
                non_allocating: 1
                com.delphix:indirect_object: 7362
                com.delphix:indirect_births: 7364
                com.delphix:prev_indirect_vdev: 6
                create_txg: 184997
                com.delphix:vdev_zap_top: 7358
            children[9]:
                type: 'mirror'
                id: 9
                guid: 6572754825354807752
                metaslab_array: 7252
                metaslab_shift: 29
                ashift: 12
                asize: 107369463808
                is_log: 0
                create_txg: 185045
                com.delphix:vdev_zap_top: 7154
                children[0]:
                    type: 'disk'
                    id: 0
                    guid: 18412959171612156138
                    path: '/dev/vgssd1000/z_special_ssd1000'
                    whole_disk: 0
                    DTL: 8113
                    create_txg: 185045
                    com.delphix:vdev_zap_leaf: 7155
                children[1]:
                    type: 'disk'
                    id: 1
                    guid: 14761859464850297329
                    path: '/dev/vg/z_special_ssd300'
                    whole_disk: 0
                    DTL: 8112
                    create_txg: 185045
                    com.delphix:vdev_zap_leaf: 7156
        features_for_read:
            com.delphix:hole_birth
            com.delphix:embedded_data
            com.delphix:device_removal
```

</details>

### Include any warning/errors/backtraces from the system logs

None.

**Comment by rincebrain** [comment]
"Issued" not actually issuing reads to the data vdevs is very surprising, to put it mildly.

You said you made the pool on 2.1.5 - did you migrate from 2.1.5 for some reason which would prevent you going back to 2.1.5 and seeing if it scrubs correctly?

@amotin has a number of performance enhancements touching scrub in the last little while in git, maybe try rolling back to before some of those and seeing if it still does that? (Maybe hop back to e.g. 87b46d63b283d9e2c3b10945f37233d1dedef02a and see if that has this behavior, then bisect from there?)

**Comment by speed47** [comment]
I'm under 2.1.99 because I'm chasing another bug, I can totally go back to 2.1.5 as I have not upgraded the pool since creation. I'll actually try to go back as far as I can and the feature bits permit, as long as I can reproduce the issue. Let's bisect indeed!

**Comment by speed47** [comment]
I can reproduce it even with OpenZFS 2.0.0 (I had to downgrade to a 5.9 kernel to test that). So, recent commits around scrub don't seem to be the cause. I can't go further in the past as my zpool uses zstd compression.

**Comment by rincebrain** [comment]
...very strange. Let me see what I can find, though my usual testbed for spinning rust is currently under the weather as it had to be emergency repurposed...

**Comment by speed47** [comment]
I can share my zpool history if it helps. You'll notice there were a bunch of devices add/remove/attach/detach, this is due to me migrating from btrfs to zfs with a careful disk rotation to never go under a mirror-like redundancy, and then hitting another zfs bug (a possible corruption one this time), and trying too stabilize my pool, which is OK now. I'm currently finalizing a reproducer for that other bug, but in the meantime, yep, I'd have liked very much to be able to scrub my pool, as it's not a test one... (even if I do have backups of my important data, more backups than a sane person would probably have, actually)

<details>
<summary>zpool history z | grep -w zpool | grep -vw -e scrub -e import -e load-key</summary>

```
2022-07-21.01:04:57 zpool create -f -o ashift=12 -o feature@encryption=enabled z mirror /dev/disk/by-id/usb-WD_Elements_25A3_35514747344C4242-0:0 /dev/disk/by-id/ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F
2022-07-21.01:05:25 zpool add -f z mirror /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR
2022-07-21.01:07:15 zpool add z special /dev/mapper/vgssd120-z_special /dev/mapper/vg-z_special
2022-07-23.17:49:06 zpool export z
2022-07-23.17:49:32 zpool export z
2022-07-23.17:56:49 zpool add z special /tank/SPECIAL
2022-07-23.18:01:20 zpool remove z vg-z_special
2022-07-23.18:06:43 zpool add z special mirror /dev/vg/z_special /dev/vgssd120/z_special
2022-07-23.18:06:56 zpool remove z /tank/SPECIAL
2022-07-28.16:00:01 zpool detach z /dev/vgssd120/z_special
2022-07-28.16:00:40 zpool attach z /dev/vg/z_special /dev/vgssd1000/z_specialtmp
2022-07-28.16:29:12 zpool clear z /dev/vgssd1000/z_specialtmp <== started to hit the other bug I'll post later (apparently wrong CKSUM fails)
2022-07-28.16:34:32 zpool replace z /dev/vgssd1000/z_specialtmp /dev/disk/by-id/ata-INTEL_SSDSA2CW120G3_CVPR1475006P120LGN
2022-07-28.16:51:54 zpool attach z /dev/vg/z_special /dev/vgssd1000/z_special_ssd1000
2022-07-28.17:23:43 zpool clear z ata-INTEL_SSDSA2CW120G3_CVPR1475006P120LGN
2022-07-28.17:24:08 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.17:31:16 zpool detach z ata-INTEL_SSDSA2CW120G3_CVPR1475006P120LGN
2022-07-28.18:10:03 zpool attach z /dev/vg/z_special /dev/vgssd300/z_special_ssd300
2022-07-28.18:54:14 zpool detach z 3399941838729346622
2022-07-28.18:54:27 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.19:16:30 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.19:23:01 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.19:23:18 zpool attach z /dev/vg/z_special /z/tank/SPECIAL
2022-07-28.19:39:11 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.19:39:19 zpool clear z /z/tank/SPECIAL
2022-07-28.20:42:53 zpool detach z 6808301552009113104
2022-07-28.20:43:47 zpool attach z /dev/vg/z_special /dev/vgssd1000/z_special_ssd1000b
2022-07-28.20:53:48 zpool clear z /dev/vgssd1000/z_special_ssd1000
2022-07-28.22:46:39 zpool clear z
2022-07-28.22:54:05 zpool detach z /dev/vgssd1000/z_special_ssd1000
2022-07-28.22:54:07 zpool detach z /dev/vgssd1000/z_special_ssd1000b
2022-07-28.22:56:07 zpool attach z /dev/vg/z_special /dev/vgssd1000/z_special_ssd1000
2022-07-28.23:35:06 zpool add z special mirror /dev/vgssd1000/z_special2_ssd1000a /dev/vgssd1000/z_special2_ssd1000b
2022-07-28.23:35:39 zpool detach z /dev/vgssd1000/z_special_ssd1000
2022-07-28.23:36:24 zpool remove z /dev/vg/z_special
2022-07-28.23:50:07 zpool attach z /dev/vgssd1000/z_special2_ssd1000a /dev/vg/z_special_ssd300
2022-07-29.09:22:44 zpool attach z /dev/vgssd1000/z_special2_ssd1000a /dev/vgssd1000/z_special2_ssd1000c
2022-07-29.09:43:51 zpool attach z /dev/vgssd1000/z_special2_ssd1000a /dev/vgssd1000/z_special2_ssd1000d
2022-07-29.10:05:19 zpool clear z
2022-07-29.10:09:31 zpool detach z /dev/vgssd1000/z_special2_ssd1000c
2022-07-29.10:09:32 zpool detach z /dev/vgssd1000/z_special2_ssd1000d
2022-07-29.10:09:43 zpool add z special mirror /dev/vgssd1000/z_special2_ssd1000c /dev/vgssd1000/z_special2_ssd1000d
2022-07-29.10:10:23 zpool remove z mirror-6
2022-07-29.10:37:43 zpool attach z /dev/vgssd1000/z_special2_ssd1000c /dev/vg/z_special_ssd300
2022-07-29.10:52:48 zpool detach z /dev/vg/z_special_ssd300
2022-07-29.10:53:10 zpool add z special mirror /dev/vgssd1000/z_special2_ssd1000a /dev/vg/z_special_ssd300
2022-07-29.10:53:28 zpool remove z mirror-8
2022-07-29.10:54:39 zpool add z special mirror /dev/vgssd1000/z_special_ssd1000 /dev/vg/z_special_ssd300
2022-07-29.10:54:49 zpool remove z mirror-7

/dev/vgssd1000/z_special_ssd1000* : 100G LVM
/dev/vgssd300/z_special_ssd300: 100G LVM
/dev/vgssd120/z_special /dev/mapper/vg-z_special: 2x50G LVM
/dev/vgssd1000/z_specialtmp: 50G LVM
/tank/SPECIAL: 100G file on another filesystem
/z/tank/SPECIAL: 100G file on the pool itself (stupid idea, but I had to try it)
/dev/disk/by-id/usb-WD_Elements_25A3_35514747344C4242-0:0 /dev/disk/by-id/ata-WDC_WD120EDAZ-11F3RA0_5PG63E9F: 2x12T spinning rust
/dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2DVRDXF /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N2STUCVR 2x3T spinning rust
```

</details>

Not sure you really want to try to reproduce that exactly, but it may give you a hint

**Comment by rincebrain** [comment]
If reproducing it exactly works, then that's fine, I can do it and then prune it.

That said, while I'm looking into that, I should ask about the least surprising conclusion - you don't have the tunable `zfs_no_scrub_io` set, do you?

**Comment by speed47** [comment]
Oh well, I think we're going to be extremely non-surprised then:
```
# cat /sys/module/zfs/parameters/zfs_no_scrub_io
1

# cat /etc/modprobe.d/zfs.conf
options zfs zfs_arc_max=0x18000000 zfs_no_scrub_io=1 zfs_no_scrub_prefetch=1 zfs_scrub_delay=100000000 l2arc_noprefetch=1 zfs_prefetch_disable=1

# l /etc/modprobe.d/zfs.conf
-rw-r--r-- 1 root root 146 Jun 15  2012 /etc/modprobe.d/zfs.conf

```

I'll just remove this file, as the defaults are probably way saner that whatever I was trying to do 10 years ago (and ended up not using zfs on this system up until last week, so never noticed).

If confirmation was still needed, the ETA is way more aligned with that I would expect on this zpool now that all the tunables are back to default:
```
  pool: z
 state: ONLINE
  scan: scrub in progress since Sat Jul 30 16:44:12 2022
        758G scanned at 15.8G/s, 9.30G issued at 198M/s, 12.9T total
        0B repaired, 0.07% done, 18:51:46 to go
```

Thanks for your hint. I would probably be out of hair before finally finding out the cause...

**Comment by speed47** [comment]
Wait, now I can't reproduce the other bug I have when `zfs_no_scrub_io` is set to 0!
I think it's an unwanted side effect of this tunable. My reproducer worked on the original machine but not in a disposable VM on another machine, however now that I tried to set this tunable to 1 on the VM, I can reproduce it immediately on both systems.

Before I open another issue, do we agree that setting this tunable to 1 should never create checksum events in `zpool events` and having the CKSUM counter go up on one of the devices of a mirror vdev, using a precise sequence of zpool add/remove/attach commands?

**Comment by rincebrain** [comment]
That sounds like an accurate statement to my incomplete knowledge, yes.

**Comment by speed47** [comment]
OK, I'll create another issue for this and we'll see where it brings us.
Thanks for your help on this one in any case!
